compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/LargeMethodSplitter.java

```diff
@@ -0,0 +1,379 @@
+                    (((BLangAssignment) stmt).varRef instanceof BLangSimpleVarRef.BLangPackageVarRef) &&
+                    Symbols.isFlagOn(
+                            ((BLangSimpleVarRef.BLangPackageVarRef) ((BLangAssignment) stmt).varRef).varSymbol.flags,
+                            Flags.LISTENER)
+            ) {
+                // this is where listener registrations starts, they are independent stmts
+                break;
+            }
+        }
+        newFuncBody.stmts.addAll(chunkStmts);
+
+        // rest of the statements can be split without chunks
+        for (int i = varDefIndex; i < stmts.size(); i++) {
+            if (i > 0 && i % methodSize == 0) {
+                generatedFunctions.add(newFunc);
+                newFunc = createIntermediateInitFunction(packageNode, env);
+                newFuncBody = (BLangBlockFunctionBody) newFunc.body;
+                symTable.rootScope.define(names.fromIdNode(newFunc.name), newFunc.symbol);
+            }
+            newFuncBody.stmts.add(stmts.get(i));
+        }
+
+        generatedFunctions.add(newFunc);
+
+        for (int j = 0; j < generatedFunctions.size() - 1; j++) {
+            BLangFunction thisFunction = generatedFunctions.get(j);
+
+            BLangCheckedExpr checkedExpr =
+                    ASTBuilderUtil.createCheckExpr(initFunction.pos,
+                            desugar.createInvocationNode(generatedFunctions.get(j + 1).name.value,
+                                    new ArrayList<>(), symTable.errorOrNilType),
+                            symTable.nilType);
+            checkedExpr.equivalentErrorTypeList.add(symTable.errorType);
+
+            BLangExpressionStmt expressionStmt = ASTBuilderUtil
+                    .createExpressionStmt(thisFunction.pos, (BLangBlockFunctionBody) thisFunction.body);
+            expressionStmt.expr = checkedExpr;
+            expressionStmt.expr.pos = initFunction.pos;
+
+            if (j > 0) { // skip init func
+                thisFunction = desugar.rewrite(thisFunction, env);
+                packageNode.functions.add(thisFunction);
+                packageNode.topLevelNodes.add(thisFunction);
+            }
+        }
+
+        rewriteLastSplitFunction(packageNode, env, generatedFunctions);
+        initFuncIndex = 0;
+        return generatedFunctions.get(0);
+    }
+
+    private boolean isAssignmentWithInitOrRecordLiteralExpr(BLangStatement statement) {
+        if (statement.getKind() == NodeKind.ASSIGNMENT) {
+            return desugar.isMappingOrObjectConstructorOrObjInit(((BLangAssignment) statement).getExpression());
+        }
+        return false;
+    }
+
+    /**
+     * Split package start function into several smaller functions.
+     *
+     * @param packageNode package node
+     * @param env symbol environment
+     * @return initial start function but trimmed in size
+     */
+    private BLangFunction splitStartFunction(BLangPackage packageNode, SymbolEnv env) {
```

Comment:
Shall we add a comment with how function is splitted.

Reply:
Added comments.